File: leetcode/problems_list/242-ValidAnagram/solution.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function (s, t) {
  // Hash Table Solution / Time Complexity O(n) / Space Complexity O(n)
  let sWord = {};
  let tWord = {};
  if (s.length !== t.length) return false;
  for (let i = 0; i < s.length; i++) {
    sWord[s[i]] = sWord[s[i]] ? sWord[s[i]] + 1 : 1;
    tWord[t[i]] = tWord[t[i]] ? tWord[t[i]] + 1 : 1;
  }
  for (let key in sWord) {
    if (sWord[key] !== tWord[key]) return false;
  }
  return true;
};
/** Sorted Array Solution / Time Complexity O(nlogn) / Space Complexity O(n)
 * let [sWord, tWord] = [[...s].sort(), [...t].sort()]
 * let anagramCheck = true
 *   if (sWord.length !== tWord.length) return !anagramCheck;
 * for (let i = 0; i < sWord.length; i++) {
 * if (sWord[i] !== tWord[i]) anagramCheck = false;
 * }
 * return anagramCheck;
 */

File: leetcode/problems_list/242-ValidAnagram/solution.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        count = {}
        for curr in s:
            count[curr] = count.get(curr, 0) + 1
        for curr in t:
            count[curr] = count.get(curr, 0) - 1
        for val in count.values():
            if val != 0:
                return False
        return True


"""
# Sorted Array Solution Time Complexity O(nlogn) Space Complexity O(n)
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
      sWord = sorted(list(s))
      tWord = sorted(list(t))
      anagramCheck = True
      if (len(sWord) != len(tWord)):
          return not anagramCheck
      for i in range(len(sWord)):
        if (sWord[i] != tWord[i]):
          anagramCheck = False
      return anagramCheck
"""
